setTimeout to allow closing Link UI on second click

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
We need a second click on the same link to close the link popover

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Otherwise, you can't place the caret within the editor to edit link text. The popover will always steal focus.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Add a clickDelay via a 100ms setTimeout to prevent the link control ui from opening again. 

We can't toggle the `addingLink` state via the link because the `onFocusOutside` event fires to close the popover, and a the click on the inline link opens the link control again. A better fix to this issue would be to have the `onFocusOutside` event report the correct element that was clicked, and we could check the element that was clicked. Since we can't do that, the current flow looks like:

- Click a link: `setAddingLink( true )`
- Click the same link again:
  -  `onFocusOutside`: `setAddingLink( false )`
  - link click handler sets it back to true: `setAddingLink( true )`

Flow from this PR:
- Click a link: `setAddingLink( true )`
- Store the element that was clicked: `opendBy.current = event.target`
- Click the same link again:
  - `onFocusOutside`: `setAddingLink( false )`
  - Set a timeout: `clickTimeoutId.current = setTimeout( () => { clickTimeout.current = false; }, 100 );`
  - link click handler returns early if `clickTimeout.current` and `opendBy` is the same element that was clicked.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Add several links to a paragraph
- Click between them 
- Double click them
- Use the toolbar to open the link control
- Use the keyboard to escape from it
- Mix all together and try to break it.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

scruffian commented:
This works well until I try to move between two different links:
https://github.com/WordPress/gutenberg/assets/275961/536d10f9-78b1-4c21-80f5-2add595164f3

getdave commented:
Thanks for attempting another route. I am slightly concerned that adding more "hacks" (your words not mine 😅) could compound what is already a complex set of interactions. 

It's good to have this exploration though 🙇

scruffian commented:
This is now testing well for me.

One thing we could do to mitigate the risk of having so much code that we are unhappy with, is to make a PR that removes all the "hacks", so at least we have them documented, and then once the underlying issues are fixed we can merge that PR...

jeryj commented:
> I am slightly concerned that adding more "hacks"

Me too. I don't see a way around not adding a `setTimeout` in this case to "skip" the `onFocusOutside` handler until `onFocusOutside` can accurately report the element that was clicked.

However, this isn't the most egregious hack. It's definitely not "the react way" though.

draganescu commented:
Applying this patch:

[count-mouse-down.patch](https://github.com/WordPress/gutenberg/files/14523848/count-mouse-down.patch)

1. What is broken by counting to one and back to zero instead of using a timeout?
2. Using mouseDown instead of click, we now also allow the user to not only click into the text but also to directly select a few chars and have the selection remain.

jeryj commented:
> What is broken by counting to one and back to zero instead of using a timeout?

The main one is that we have to have a way to reset it to zero to allow a click on the link or toolbar button to open the popover:

- Click link
- Popover opens
- Click text in the same paragraph
- onFocusOutside sets clickCount.current to 1
- click link again
- clickCount.current is still 1, as it hasn't been reset, so link popover doesn't open

We'd need a timeout to reset the clickCount to 0 which brings us back to relying on a timeout 😞

jeryj commented:
> One thing we could do to mitigate the risk of having so much code that we are unhappy with, is to make a PR that removes all the "hacks", so at least we have them documented, and then once the underlying issues are fixed we can merge that PR...

https://github.com/WordPress/gutenberg/pull/59713

getdave commented:
I think we can close this one now that https://github.com/WordPress/gutenberg/pull/59635 has merged. I'll leave it up to you @jeryj

jeryj commented:
This still does add a feature that is otherwise missing (second click on the link button or link text to close the popover). We don't need to cherry-pick it, but I think it will work in tandem with https://github.com/WordPress/gutenberg/pull/59635/ after rebasing and fixing. I'd like to leave it open to discuss if it's still a worthwhile feature.

jeryj committed:
Add clarifying comment

jeryj committed:
Add a clickDelay via a 100ms setTimeout to prevent the link control ui from opening again.

We can't toggle the addingLink state via the link because the onFocusOutside event fires to close the popover, and a the click on the inline link opens the link control again. A better fix to this issue would be to have the onFocusOutside event report the correct element that was clicked, and we could check the element that was clicked.

draganescu commented:
I don’t think a second click to close the popover for links is a worthwhile feature.

Now that `trunk` allows for caret placement inside the link format and for text to be selected inside the link format, the fact that the link popover is open while performing actions on the formatted text is a good thing, as it's an inseparable part of the UI signalling of this format in particular (link).

Also other formats have their UI active all the time (e.g. bold) it doesn’t disappear.
